### PR TITLE
Implementation of KEP-3589: uniformly filter manageJobsWithoutQueueNames by namespace

### PR DESCRIPTION
**Pull request by dgrove-oss**
#### What type of PR is this?

/feature

#### What this PR does / why we need it:

Implements core feature of KEP-3589 by filtering default suspension of newly created jobs by a namespace selector.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Adds a managedJobsNamespaceSelector to the Kueue configuration that enables namespace-based control of whether Jobs submitted without a `kueue.x-k8s.io/queue-name` label are managed by Kueue for all supported Job Kinds.
```

### CODE REVIEW

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -180,6 +187,7 @@ var defaultFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
 	ExposeFlavorsInLocalQueue:           {Default: true, PreRelease: featuregate.Beta},
 	AdmissionCheckValidationRules:       {Default: false, PreRelease: featuregate.Deprecated},
 	KeepQuotaForProvReqRetry:            {Default: false, PreRelease: featuregate.Deprecated},
+	ManagedJobsNamespaceSelector:        {Default: true, PreRelease: featuregate.Beta},
```
Comment (mbobrovskyi) [inline pkg/features/kube_features.go:190]: Please also update [The currently supported features table](https://github.com/kubernetes-sigs/kueue/blob/dff92c9bcb5411d2b12bbf3240c977361d0ef87d/site/content/en/docs/installation/_index.md?plain=1#L234)

File: pkg/controller/jobframework/defaults.go [hunk 2]
```diff
@@ -17,18 +17,43 @@ limitations under the License.
...
 	// Do not default suspend a job whose owner is already managed by Kueue
 	if owner := metav1.GetControllerOf(job.Object()); owner != nil && IsOwnerManagedByKueue(owner) {
-		return
+		return nil
+	}
+
+	// Do not default suspend a job without a queue name unless the namespace selector also matches
+	if QueueName(job) == "" && manageJobsWithoutQueueName && managedJobsNamespaceSelector != nil {
+		ns := corev1.Namespace{}
+		err := k8sClient.Get(ctx, client.ObjectKey{Name: job.Object().GetNamespace()}, &ns)
```
[the remark this answers is not in the document]
Comment (dgrove-oss) [reply pkg/controller/jobframework/defaults.go:40]: Should be using the cache.  Similar get is already in the pod webhook.

File: pkg/features/kube_features.go (hunk 1 above)
Comment (dgrove-oss) [reply pkg/features/kube_features.go:190]: Table updated.

File: pkg/controller/jobframework/defaults.go [hunk 3]
```diff
@@ -17,18 +17,43 @@ limitations under the License.
...
+		if err != nil {
+			return fmt.Errorf("failed to get namespace: %w", err)
+		}
+		nsSelector, err := metav1.LabelSelectorAsSelector(managedJobsNamespaceSelector)
+		if err != nil {
+			return fmt.Errorf("failed to parse namespace selector: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (dgrove-oss) [reply pkg/controller/jobframework/defaults.go]: Yes.  I reworked to parse selector during startup and store the resulting `labels.Selector` in the webhooks/job reconciler.

File: pkg/config/validation.go [hunk 4]
```diff
@@ -205,31 +206,42 @@ func validatePodIntegrationOptions(c *configapi.Configuration) field.ErrorList {
...
-	if err != nil {
-		return allErrs
-	}
+		for _, pn := range prohibitedNamespaces {
+			if selector.Matches(pn) {
+				allErrs = append(allErrs, field.Invalid(managedJobsNamespaceSelectorPath, c.ManagedJobsNamespaceSelector,
+					fmt.Sprintf("should not match the %q namespace", pn[corev1.LabelMetadataName])))
+			}
+		}
+	} else if c.Integrations.PodOptions != nil && c.Integrations.PodOptions.NamespaceSelector != nil {
```
[the remark this answers is not in the document]
Comment (dgrove-oss) [reply pkg/config/validation.go]: reworked this.

**Comment by dgrove-oss** [comment]
unit and existing integration tests now passing.  

I don't quite understand the `kueue-verify-main` error is trying to tell me.  Any hints?
```/home/prow/go/src/sigs.k8s.io/kueue/bin/golangci-lint run --timeout 15m0s
pkg/controller/jobframework/defaults.go:26: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(sigs.k8s.io/kueue) -s blank -s dot (gci)
	"sigs.k8s.io/controller-runtime/pkg/client"
```

ideally, we'd add additional integration tests that verified the new feature across all the integrations, but I ran out of time today.  Could wait for a followup PR if necessary due to 0.10 deadline.

File: pkg/controller/jobframework/defaults.go [hunk 5]
```diff
@@ -17,18 +17,39 @@ limitations under the License.
...
 
 import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kueue/pkg/features"
```
Comment (tenzen-y) [inline pkg/controller/jobframework/defaults.go:27]:
```suggestion
	"sigs.k8s.io/controller-runtime/pkg/client"

	"sigs.k8s.io/kueue/pkg/features"
```

This should resolve CI errors.

**Review by tenzen-y — COMMENTED** [review]
I did not review the entire codes. I just left a comment how to resolve CI errors.

**Comment by mimowo** [comment]
> ideally, we'd add additional integration tests that verified the new feature across all the integrations, but I ran out of time today. Could wait for a followup PR if necessary due to 0.10 deadline.

I think we are good to move the release for tomorrow or EOW. I

File: pkg/controller/jobframework/defaults.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/jobframework/defaults.go:41]: sgtm

**Comment by mimowo** [comment]
/lgtm
/approve
Please add "Part of #3589" in the description. As for a follow up, I think we should respect the new selector in webhooks for deployment and StatefulSet, before we set the PodGroup labels and annotations. Feel free to follow up within the issue.

**Comment by tenzen-y** [comment]
/release-note-edit
```release-note
ConfigAPI: Adds a managedJobsNamespaceSelector to the Kueue configuration that enables namespace-based control of whether Jobs submitted without a `kueue.x-k8s.io/queue-name` label are managed by Kueue for all supported Job Kinds.
```

**Comment by tenzen-y** [comment]
/release-note-edit
```release-note
Adds a managedJobsNamespaceSelector to the Kueue configuration that enables namespace-based control of whether Jobs submitted without a `kueue.x-k8s.io/queue-name` label are managed by Kueue for all supported Job Kinds.
```

**Pull request merged**